### DISCUSSION TITLE
Compiling on Windows

### DISCUSSION THREAD

**Discussion by voltagex** [mlin/sqlite_zstd_vfs#10] [Q&A]
I got as far as 

* Installing VC++ build tools from https://download.visualstudio.microsoft.com/download/pr/949751db-6687-4a88-a0cf-047f10908a29/fd92768eaad0531e50e0010414d304675da7b859b138d865127476f22b79e406/vs_BuildTools.exe
* Installing vcpkg and sqlite3 (headers)
* Building using cmake -DCMAKE_CXX_FLAGS="/DWIN32 /D_WINDOWS /W3 /GR /EHsc /I C:\Users\WDAGUtilityAccount\vcpkg\installed\x86-windows\include" -B build / cmake --build

(ignore the weird path, was testing in Windows Sandbox to make cleanup easier)

However, it looks like there's a dependency on a posix header called 'libgen.h'. I'm wondering how much I'd have to change without needing to go over to MinGW.

**Comment by mlin** [comment]
Hi! In general, I've not made any attempt to build/test this on Windows (and I don't use Windows in my day-to-day flow) so I welcome any efforts that can be spared to that end. Ideally we'd add a Windows job to [the GitHub Actions build](https://github.com/mlin/sqlite_zstd_vfs/blob/main/.github/workflows/tests.yml).

However, [the libgen.h include was spurious and I also ifdef'ed-out a use of unistd.h](https://github.com/mlin/sqlite_zstd_vfs/commit/b2a443bbe09c25c5c50bf7820b2a2167920b7076). Possibly this might get you to the next problem =)

**Reply by voltagex** [reply]
Hit the next problem

```
CMake Error at C:/vcpkg/scripts/buildsystems/vcpkg.cmake:148 (if):
  Maximum recursion depth of 1000 exceeded
Call Stack (most recent call first):
  C:/vcpkg/scripts/buildsystems/vcpkg.cmake:605 (z_vcpkg_function_arguments)
 ...snipped about a hundred repeated C:/vcpkg/scripts/buildsystems/vcpkg.cmake:606 (_add_library) lines...
  build/_deps/concurrentqueue-src/CMakeLists.txt:6 (add_library)

```

CMakeLists.txt:6 in concurrentqueue is `add_library(${PROJECT_NAME} INTERFACE)` which is probably where the recursion is coming from

I know vcpkg isn't your problem, but I'd really like to use it on Windows as it makes a lot of things simpler.

(Deleted previous post about FetchContent being the issue, red herring.)

**Reply by voltagex** [reply]
I think this might be a bit beyond my CMake abilities.

I got as far as [CMakeLists.txt](https://github.com/mlin/sqlite_zstd_vfs/files/8580820/CMakeLists.txt) which was configured via `cmake -DVCPKG_TARGET_TRIPLET=x64-windows -DCMAKE_TOOLCHAIN_FILE=C:\vcpkg\scripts\buildsystems\vcpkg.cmake -B build` and build with `cmake --build build`


```
C:\git\sqlite_zstd_vfs\src\SQLiteNestedVFS.h(25,10): fatal error C1083: Cannot open include file: 'SQLiteVFS.h': No suc
h file or directory [C:\git\sqlite_zstd_vfs\build\nested_vfs.vcxproj]
  Building Custom Rule C:/git/sqlite_zstd_vfs/CMakeLists.txt
  zstd_vfs.cc
C:\git\sqlite_zstd_vfs\src\zstd_vfs.cc(6,10): fatal error C1083: Cannot open include file: 'web_vfs.h': No such file or
 directory [C:\git\sqlite_zstd_vfs\build\zstd_vfs.vcxproj]
```

vcpkg was configured via `vcpkg install sqlitecpp:x64-windows zstd:x64-windows sqlite3:x64-windows`

I *think* making `sqlite_web_vfs` optional, using upstream/vcpkg sqlitecpp and then some kind of fix to the include directories would get me there, but I think at this point I've hacked up your project too much for the sake of a platform you don't use.

**Comment by voltagex** [comment]
I won, but I don't think I'd do it again.

See https://github.com/mlin/sqlite_zstd_vfs/compare/main...voltagex-forks:feature/windows-hacks-1?expand=1

I would definitely not expect you to carry this patch.

This will likely work for what I need it to do, and I hope someone with more Windows and CMake expertise can use this as a starting point in the future.

```
C:\git\sqlite_zstd_vfs\build\Debug [feature/windows-hacks-1 +0 ~3 -0 !]> sqlite3 Chinook_Sqlite.sqlite
SQLite version 3.37.1 2021-12-30 15:30:28
Enter ".help" for usage hints.
sqlite> .load zstd_vfs
sqlite> VACUUM INTO 'file:Chinook_Sqlite.zstd.sqlite?vfs=zstd'
   ...> ;
C:\git\sqlite_zstd_vfs\src\SQLiteNestedVFS.h:1244: file:C%3A%5Cgit%5Csqlite_zstd_vfs%5Cbuild%5CDebug%5CChinook_Sqlite.zstd.sqlite
WARNING: The maximum dictionary size 98304 is too large compared to the source size 76793! size(source)/size(dictionary) = 0.781179, but it should be >= 10! This may lead to a subpar dictionary! We recommend training on sources at least 10x, and preferably 100x the size of the dictionary!
C:\git\sqlite_zstd_vfs\src\zstd_vfs.h:273: dict 1 @ 384 264ms
C:\git\sqlite_zstd_vfs\src\SQLiteNestedVFS.h:1053: reads sequential: 26 non-sequential: 74 interior: 7 longest: 0 prefetched: 0 wasted: 0 seek: 0ms decode: 2ms
sqlite>
```

Thanks for your time and the libgen fix.

**Reply by mlin** [reply]
Thanks for reporting back! Maybe I'll get back to Windows dev one day, until then, I'd definitely welcome anyone picking this up!

**Comment by friendlyanon** [comment]
The issue here is twofold:

- This project forcibly vendors its dependencies.
- The dependencies do not have a CMake package.

First order of business would be fixing the dependencies, so they provide proper CMake packages that can be used with `find_package()` and then convert this project to use `find_package()` as well.

If @mlin wishes not to mess with vcpkg's overlay ports to adapt those dependencies, the dependencies can still be vendored for dev workflows using find modules like so: https://github.com/friendlyanon/cmake-init-fetchcontent  
That example repository shows how to use the idiomatic way of importing dependencies (`find_package()`), while still vendoring the dependencies under the hood.

I'd gladly review PRs to this project and its dependencies to ensure the CMake code is proper if I'm pinged and the above example is not enough as guidance.

**Comment by ronniec95** [comment]
Thanks for this build. It's almost working, except right at the end

    Project "D:\sqlite_zstd_vfs-174e2a03f6dbe9c0d94245e461866d1178a69d29\sqlite_zstd_vfs-174e2a03f6dbe9c0d94245e461866d1178a69d29\CMakeFiles\3.26.0-msvc3\VCTargetsPath.vcxproj" on node 1 (default targets).
    C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(832,5): error : **The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.** 

I think you need to add dependencies to readerwriterqueue, sqlitevfs; and it also looks like SQLiteCPP needs to have declspec(dllexport) (vcpkg?) defined for it's types otherwise you get a missing symbols link error. That's as far as I've got but still going through the build.
